drivers/platform/stm32/stm32_irq.c

```diff
@@ -0,0 +1,264 @@
+
+	return 0;
+}
+
+/**
+ * @brief Not implemented. On stm32 platform, trigger level is set in the app, prior to irq_* configuration.
+ * @param desc - gpio irq descriptor.
+ * @param irq_id - the pin number
+ * @param level - the trigger condition.
+ * @return -ENOSYS
+ */
+int32_t stm32_trigger_level_set(struct irq_ctrl_desc *desc,
+				uint32_t irq_id,
+				enum irq_trig_level level)
+{
+	int ret;
+	EXTI_ConfigTypeDef config;
+	ret = HAL_EXTI_GetConfigLine(desc->extra, &config);
+	if (ret != HAL_OK)
+		return -EFAULT;
+
+	switch (level) {
+	case IRQ_EDGE_FALLING:
+		config.Trigger = EXTI_TRIGGER_FALLING;
+		break;
+	case IRQ_EDGE_RISING:
+		config.Trigger = EXTI_TRIGGER_RISING;
+		break;
+	case IRQ_EDGE_BOTH:
+		config.Trigger = EXTI_TRIGGER_RISING_FALLING;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = HAL_EXTI_SetConfigLine(desc->extra, &config);
+	if (ret != HAL_OK)
+		return -EFAULT;
+
+	return 0;
+}
+
+/**
+ * @brief Not implemented.
+ * @param desc - The IRQ controller descriptor.
+ * @param irq_id - Interrupt identifier.
+ * @param callback - Descriptor of the callback.
+ * @return -ENOSYS
+ */
+int32_t stm32_irq_register_callback(struct irq_ctrl_desc *desc, uint32_t irq_id,
+				    void *callback)
+{
+	int ret;
+
+	switch (desc->source) {
+	case NO_OS_GPIO_IRQ:
+		ret = HAL_EXTI_RegisterCallback(desc->extra,
+						((struct stm32_callback *)callback)->event,
+						((struct stm32_callback *)callback)->callback);
+		if (ret != HAL_OK) {
+			ret = -EFAULT;
+			break;
+		}
+		hexti[((EXTI_HandleTypeDef *)desc->extra)->Line & 0xff] = desc->extra;
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Not implemented
+ * @param desc - Interrupt controller descriptor.
+ * @param irq_id - Id of the interrupt
+ * @param callback - Descriptor of the callback.
+ * @return -ENOSYS
+ */
+int32_t stm32_irq_unregister_callback(struct irq_ctrl_desc *desc,
+				      uint32_t irq_id, void *callback)
+{
+	int ret;
+
+	switch (desc->source) {
+	case NO_OS_GPIO_IRQ:
+		hexti[((EXTI_HandleTypeDef *)desc->extra)->Line & 0xff] = NULL;
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Enable all interrupts
+ * @param desc - Interrupt controller descriptor.
+ * @return 0
+ */
+int32_t stm32_irq_global_enable(struct irq_ctrl_desc *desc)
+{
+	__enable_irq();
+
+	return 0;
+}
+
+/**
+ * @brief Disable all interrupts
+ * @param desc - Interrupt controller descriptor.
+ * @return 0
+ */
+int32_t stm32_irq_global_disable(struct irq_ctrl_desc *desc)
+{
+	__disable_irq();
+
+	return 0;
+}
+
+/**
+ * @brief Enable a specific interrupt
+ * @param desc - Interrupt controller descriptor.
+ * @param irq_id - Interrupt identifier
+ * @return 0 in case of success, errno error codes otherwise.
+ */
+int32_t stm32_irq_enable(struct irq_ctrl_desc *desc, uint32_t irq_id)
+{
+	// TODO: add a proper priority setting function in irq.h instead of this default
+	HAL_NVIC_SetPriority(irq_id, 0, 0);
```

Comment:
Adding a priority setting API seems like a good idea to me. We shouldn't do it here.

Reply:
True, but I view this as a separate topic. This PR is quite controversial already so perhaps we should do it separately. The code will work anyway with all interrupts of (prio 0, subprio 0 = highest) priority untill we do that.